elasticsearch/src/main/scala/akka/stream/alpakka/elasticsearch/ElasticsearchSourceStage.scala

```diff
@@ -0,0 +1,124 @@
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new ElasticsearchSourceLogic[T](indexName, typeName, query, client, settings, out, shape)
+
+}
+
+sealed class ElasticsearchSourceLogic[T](indexName: String,
+                                         typeName: String,
+                                         query: String,
+                                         client: RestClient,
+                                         settings: ElasticsearchSourceSettings,
+                                         out: Outlet[OutgoingMessage[T]],
+                                         shape: SourceShape[OutgoingMessage[T]])(implicit reader: JsonReader[T])
+    extends GraphStageLogic(shape)
+    with ResponseListener {
+
+  private var started = false
+  private var scrollId: String = null
+  private val responseHandler = getAsyncCallback[Response](handleResponse)
+  private val failureHandler = getAsyncCallback[Throwable](handleFailure)
+
+  def sendScrollScanRequest(): Unit =
+    try {
+      if (scrollId == null) {
+        client.performRequestAsync(
+          "POST",
+          s"$indexName/$typeName/_search",
+          Map("scroll" -> "5m", "sort" -> "_doc").asJava,
+          new StringEntity(s"""{"size": ${settings.bufferSize}, "query": ${query}}"""),
+          this
+        )
+      } else {
+        client.performRequestAsync(
+          "POST",
+          s"/_search/scroll",
+          Map[String, String]().asJava,
+          new StringEntity(Map("scroll" -> "5m", "scroll_id" -> scrollId).toJson.toString),
+          this
+        )
+      }
+    } catch {
+      case ex: Exception => handleFailure(ex)
+    }
+
+  override def onFailure(exception: Exception) = failureHandler.invoke(exception)
+  override def onSuccess(response: Response) = responseHandler.invoke(response)
+
+  def handleFailure(ex: Throwable): Unit =
+    failStage(ex)
+
+  def handleResponse(res: Response): Unit = {
+    val json = {
+      val out = new ByteArrayOutputStream()
+      try {
+        res.getEntity.writeTo(out)
+        new String(out.toByteArray, "UTF-8")
+      } finally {
+        out.close()
+      }
+    }
+
+    val jsObj = json.parseJson.asJsObject
+
+    jsObj.fields.get("error") match {
+      case None => {
+        val hits = jsObj.fields("hits").asJsObject.fields("hits").asInstanceOf[JsArray]
+        if (hits.elements.isEmpty && scrollId != null) {
+          completeStage()
+        } else {
+          scrollId = jsObj.fields("_scroll_id").asInstanceOf[JsString].value
+          val messages = hits.elements.reverse.map { element =>
+            val doc = element.asJsObject
+            val id = doc.fields("_id").asInstanceOf[JsString].value
+            val source = doc.fields("_source").asJsObject
+            OutgoingMessage(id, source.convertTo[T])
+          }
+          emitMultiple(out, messages)
+          sendScrollScanRequest()
```

Comment:
AFAICS doing it like this would disable backpressure because there's basically a loop `sendScrollScanRequest` -> `handleResponse` -> `sendScrollScanRequest` that immediately slurps the whole data source instead of waiting until the first results are drained by the stream and a new pull comes in.

Fortunately, it should be very simple to fix. Couldn't you remove the `sendScrollScanRequest` here and also remove the condition in `onPull` below. What then should happen is that when you use `emitMultiple`, your below `OutHandler` gets exchanged with an internal `EmitMultiple` handler that will send out the current results. Once all of those results are drained your original `OutHandler` is swapped in again which will then use `sendScrollScanRequest` on the next `onPull`.